samples: subsys: mp: Add audio pipeline sample

michal-chvatal opened the pull request:
This PR introduces an audio pipeline sample built on top of libMP. The current implementation provides a simple pipeline that captures audio from a digital microphone (DMIC), applies gain control, and outputs the processed audio through an I2S codec to a speaker. Thanks to libMP’s modular design, this pipeline can be easily extended beyond its current functionality to support more complex audio processing scenarios without significantly increasing code complexity.

This PR depends on:
- https://github.com/zephyrproject-rtos/zephyr/pull/98514
- https://github.com/zephyrproject-rtos/zephyr/pull/112592
- https://github.com/zephyrproject-rtos/zephyr/pull/111526
- https://github.com/nxp-upstream/zephyr/commits/mp_aud_dependencies/ (containing various dependency commits which are upstreamed / upstreaming through different PRs, e.g. https://github.com/zephyrproject-rtos/zephyr/pull/98500)

ngphibang commented:
> This isn't necessarily a driver sample but is placed as if it is one. Perhaps mp samples should be placed in samples/lib/libmp?

In fact, it was in `samples/lib/libmp` at the beginning, then we see that libMP can be injected / leveraged into current Zephyr samples. This is true for video cases, we have 4 video samples : capture, capture-to-lvgl, tcpserversink, uvc. These samples are multimedia in general and not purely video (as they relates to other domains like display, lvgl, network, usb, etc.) but currently places in `samples/drivers/video` because the video stuffs are there.

So, with libMP, we can make a unique sample which replaces all video samples (or keep only one basic non-libMP capture sample), it fits perfectly for this purpose: https://github.com/zephyrproject-rtos/zephyr/issues/99916

That's why we want to place the libMP-based video samples at the same place at the current Zephyr video samples.
For audio, this may be not the case because it seems audio driver samples are very basic and mostly not requires complex pipeline ?

Anyways, about the folders, it seems it is better to place all libMP-based samples into `samples/lib/libmp` as you said, the only concern is if we do that, it is considered as there is [no Zephyr user of libMP](https://github.com/zephyrproject-rtos/zephyr/pull/98514#issuecomment-3556820328) and should be an external module :(.

kartben commented on samples/drivers/audio/pipeline/README.rst:
```diff
@@ -0,0 +1,94 @@
+
+    +---------------+     +---------------+     +------------- ----+     +------------------+
+    |  DMIC Source  | --> |  Caps Filter  | --> |  Gain Transform  | --> |  I2S Codec Sink  |
+    +---------------+     +---------------+     +------------------+     +------------------+
```
you should use [graphviz](https://docs.zephyrproject.org/latest/contribute/documentation/guidelines.html#graphviz) or [mermaid](https://docs.zephyrproject.org/latest/contribute/documentation/guidelines.html#mermaid) here

kartben commented on samples/drivers/audio/pipeline/README.rst:
```diff
@@ -0,0 +1,94 @@
+.. zephyr:code-sample:: audio-pipeline
+   :name: Audio pipeline
+
+      An audio pipeline built on libMP captures audio frames and applies gain control before playing them back.
```
```suggestion
   Build an audio pipeline that captures audio frames and applies gain control before playing them back, using libMP.
```

kartben reviewed (changes requested):
you should include the commits from the libmp PR so that this is tested in CI

mohittalwar23 commented:
Hello @TomasBarakNXP @ngphibang ,

I am a GSoC student under the Zephyr Project this year. My project is also focused on building a platform-agnostic audio sample. When I submitted the proposal, my plan was to write the abstraction layer myself, but after coming across libmp, my mentor @iuliana-prodan and I decided that the mediapipe subsystem is the best way forward for the sample and for plugins like i2s_src (it currently has dmic_src , also which i had to port to new subsys/mp and net_buf ).

The existing audio sample was built on the old lib/mp API. I have ported it to the new subsys/mp API, and I used your native_sim PR as the base and built on top of it. I had to remove old mp_buffer section and incorporated new structure from the different plugins. I already have a working sample here: https://github.com/mohittalwar23/zephyr/tree/gsoc/phase1-native-sim-validation

My question: should I commit into this PR, or open a new draft PR? A lot has changed in mediapipe and the original sample in this PR by @michal-chvatal is now outdated. Even my own code needs a slight refactor after moving capsfilter to the zbase plugin.

My mentor also reached out to you over email, so if possible it would be great to have a meeting.

My goal before the midterm evaluation is to get the audio sample stable on a single PR, then verify it on hardware, and from there move on to i2s_src and future work like multi-core communication and running a TinyML keyword classifier model (which was the GSoC 2025 project).

Thanks!

ngphibang commented:
@mohittalwar23: @TomasBarakNXP has rebased the audio plugin and sample to the latest MP code and will update the PRs soon so that we can work on the new code base.
We can definitely sync with a meeting.

Thanks !

mohittalwar23 commented:
Hi @ngphibang @TomasBarakNXP,

While bringing the libMP audio pipeline up on native_sim, I hit a runtime failure:

```
  <err> mp_zaud_src: Failed to get audio capabilities
  <err> mp_zaud_i2s_codec_sink: Failed to get I2S capabilities
```
 
From what I can tell, the root cause is that the get_caps API from #98500 was only implemented for the hardware drivers (dmic_mcux, dmic_nxp_micfil, i2s_mcux_flexcomm, i2s_mcux_sai, wm8904, wm8962, and da7212). The native_sim audio drivers (dmic_native_sim, i2s_native_sim, and the dummy codec codec_dummy) don't implement it, so capability negotiation fails and the pipeline aborts on native_sim.

The native_sim DMIC driver landed in #109898, which predates the get_caps API, and it looks like #98500 didn't add support there.

Did you perhaps forget to push that part, or is there another way you were testing this on native_sim? If you're busy, I'm happy to add the implementation and open a PR if that's the expected approach.

ngphibang commented:
> Did you perhaps forget to push that part, or is there another way you were testing this on native_sim? If you're busy, I'm happy to add the implementation and open a PR if that's the expected approach.

@mohittalwar23 : I believe @TomasBarakNXP forgot to add support of native_sim's caps implementation as a dependency (the last commit of this branch https://github.com/nxp-upstream/zephyr/commits/mp_aud_dependencies/). He is currently on PTO.

On his behalf, I repushed this sample PR as well as the zaud plugin PR with the latest code and without the dependency commits (which are now put separately in https://github.com/nxp-upstream/zephyr/commits/mp_aud_dependencies/), you will need to cherry-pick these commits manually. The reason to do that is the dependency commits are risked to be changed / merged along the time which are difficult to track and so, shouldn't be added together with the PR, IMO).

Sorry for this inconvenient.

mohittalwar23 commented:
Okay , No Problem. Thanks for the heads up!

ngphibang commented:
Also, convert the PR to DRAFT until all the dependencies are merged and CI passed.

josuah commented on samples/subsys/mp/dmic_i2s/src/main.c:
The tests did extra free-up of the capabilities there.

Maybe this needs some code there too? Maybe modifying libMP for having a "free" or "unref" function for when an element is done? Such as something that recursively unref all members.

ngphibang committed:
mp: Introduce Media Pipe (MP) subsystem

MP (Media Pipe) is a lightweight GStreamer-like multimedia framework
for Zephyr. MP reuses many concepts from GStreamer, such as elements,
pads, caps negotiation, and buffer negotiation and adopts a pipeline-
based architecture that decomposes multimedia processing into discrete,
interconnected elements.

It aims to simplify the development of multimedia applications by
providing simple and stable APIs for users to rapidly create their
specific applications, i.e., users simply select the built-in elements
and plugins suited to their purpose to construct a pipeline, and it
just works. This design promotes modularity, reusability, and efficient
resource management (e.g., zero-copy data flow). Moreover, the APIs
are generic enough so that application code can remain unchanged even
as MP evolves.

MP also features a highly modular, inheritance-based architecture
inspired by GStreamer, ensuring modularity, scalability, and
maintainability. For example, new custom elements can be easily added
by extending existing elements, without requiring modifications to the
core components. Plugins are decentralized from the core structures,
allowing seamless extension without altering the core framework.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>
Signed-off-by: Trung Hieu Le <trunghieu.le@nxp.com>

ngphibang committed:
mp: Add core tests

Add build-only, unit and mock pipeline tests for the MP core.

The mock pipeline is composed of a fake source, a transform and
a sink to verify the whole core framework behavior such as
pipeline creation, caps negotiation and data flow.

Assisted-by: Claude:claude-opus-4.6
Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>
Signed-off-by: Trung Hieu Le <trunghieu.le@nxp.com>

ngphibang committed:
mp: Add base plugin

Add the base plugin for the MP subsystem. This plugin provides
generic, reusable elements like:
- queue: pipeline-level threading element
- tee: pipeline branching element
- capsfilter: caps enforcement element

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

michal-chvatal committed:
mp: Add audio plugin

Add the aud (Zephyr Audio) plugin for the MP subsystem. This plugin
provides audio-specific elements that interface with Zephyr's audio
subsystems, enabling building audio capture, processing, and playback
pipelines using Zephyr audio devices.

The plugin includes audio source elements for PCM and DMIC capture,
an I2S codec sink for audio output, a gain control transform for
per-sample amplitude scaling, and audio buffer pool management.

Signed-off-by: Michal Chvatal <michal.chvatal@nxp.com>
Signed-off-by: Tomas Barak <tomas.barak@nxp.com>

michal-chvatal committed:
mp: samples: Add DMIC to I2S audio sample

Add the dmic_i2s sample application demonstrating how to build an
audio pipeline using the MP subsystem. This sample provides a simple
pipeline that captures audio from a digital microphone (DMIC), applies
gain control using the aud plugin's gain element, and outputs the
processed audio through an I2S codec to a speaker.

Signed-off-by: Michal Chvatal <michal.chvatal@nxp.com>
Signed-off-by: Tomas Barak <tomas.barak@nxp.com>

ngphibang commented on samples/subsys/mp/dmic_i2s/src/main.c:
```diff
@@ -0,0 +1,162 @@
+	default:
+		LOG_ERR("Unexpected message from element %d", msg.origin->object.id);
+		break;
+	}
```
As in other samples, the new `mp_player` utils can be used here to take control of the pipeline lifecycle and it can help to debug too (c.f. https://github.com/zephyrproject-rtos/zephyr/pull/100480/changes/c34c74bc5c0dbe2fdf62aa9b308b7e77ca302478#diff-85a3cdb502839f7f86bb5f109413e09cae1b793b01ff3e7b9ffdb80fb18aa2ccR168-R176)

mohittalwar23 replied:
Okay, I’ll pull the updates in my draft PR. Can you also review i2s src and related commits and give feedback

mohittalwar23 replied:
> As in other samples, the new `mp_player` utils can be used here to take control of the pipeline lifecycle and it can help to debug too (c.f. [c34c74b#diff-85a3cdb502839f7f86bb5f109413e09cae1b793b01ff3e7b9ffdb80fb18aa2ccR168-R176](https://github.com/zephyrproject-rtos/zephyr/commit/c34c74bc5c0dbe2fdf62aa9b308b7e77ca302478#diff-85a3cdb502839f7f86bb5f109413e09cae1b793b01ff3e7b9ffdb80fb18aa2ccR168-R176))

Hi @ngphibang thanks for the pointer to mp_player; I've adopted it in the audio_loopback sample as you suggested. It runs fine with some small warnings/errors ( which are driver side mostly )

```shell
uart:~$   player status
Player state: PLAYING
uart:~$   player stop
[00:00:36.707,000] <inf> mp_player: Player state: STOPPED
uart:~$   player play
[00:00:47.540,000] <wrn> wolfson_wm8960: target clock 4096000 can not use PLL at 12287999
[00:00:47.540,000] <wrn> wolfson_wm8960: target clock 4096000 can not use PLL at 12287999
[00:00:48.560,000] <err> mp_aud_buffer_pool: POOLDIAG rate=16000 width=32 ch=2 fi=10000 count=16 -> size=1280 min_buf=18 need=23040 budget=49152
[00:00:48.560,000] <inf> mp_aud_dmic_src: Capture started now
[00:00:48.560,000] <inf> mp_player: Player state: PLAYING
[00:00:48.595,000] <err> i2s_nxp_sai_sdma: TX underrun after 3 blocks
uart:~$ player replay
[00:01:13.034,000] <inf> mp_player: Player state: STOPPED
[00:01:13.247,000] <wrn> wolfson_wm8960: target clock 4096000 can not use PLL at 12287999
[00:01:13.247,000] <wrn> wolfson_wm8960: target clock 4096000 can not use PLL at 12287999
[00:01:14.267,000] <err> mp_aud_buffer_pool: POOLDIAG rate=16000 width=32 ch=2 fi=10000 count=16 -> size=1280 min_buf=18 need=23040 budget=49152
[00:01:14.267,000] <inf> mp_aud_dmic_src: Capture started now
[00:01:14.267,000] <inf> mp_player: Player state: PLAYING
[00:01:14.302,000] <err> i2s_nxp_sai_sdma: TX underrun after 3 blocks
uart:~$ player stop
[00:01:20.443,000] <inf> mp_player: Player state: STOPPED
uart:~$ player replay
[00:01:27.531,000] <wrn> wolfson_wm8960: target clock 4096000 can not use PLL at 12287999
[00:01:27.531,000] <wrn> wolfson_wm8960: target clock 4096000 can not use PLL at 12287999
[00:01:28.550,000] <err> mp_aud_buffer_pool: POOLDIAG rate=16000 width=32 ch=2 fi=10000 count=16 -> size=1280 min_buf=18 need=23040 budget=49152
[00:01:28.550,000] <inf> mp_aud_dmic_src: Capture started now
[00:01:28.550,000] <inf> mp_player: Player state: PLAYING
[00:01:28.585,000] <err> i2s_nxp_sai_sdma: TX underrun after 3 blocks
uart:~$ player pause
[00:01:47.267,000] <inf> mp_player: Player state: PAUSED
[00:01:47.341,000] <err> i2s_nxp_sai_sdma: TX FIFO underrun
uart:~$ player play
[00:01:59.864,000] <inf> mp_player: Player state: PLAYING
uart:~$ 


``` 
I had to do some code changes for stop to work properly.

ngphibang replied:
Are these the real bugs from the drivers and the MP's audio plugin that are revealed thanks to the mp_player ? If so, it's worth to fix them by separate PRs to upstream.
